### PR TITLE
IGNITE-22189 expiry policy info in CacheView and CacheGroupView

### PR DESCRIPTION
**Pull request by maksaska**
**What have been implemented:**
- CacheView: expiry policy factory String represantation. The CacheView#expiryPolicyFactory method returns readable output with ttl 
- CacheView: entries pending expire. The flag shows if there are entries pending to expire. The time complexity for in-memory is O(logN), and for persistance is O(Number of partitions). The flag is based on the pendingTree/pendingEntries `BPlusTree` instance, that holds the information. The instance is hold within `IgniteCacheOffheapManager` implementation class, thus the flag can only return reasonable output for caches with eager ttl (eagerTtl = true). Otherwise "Unknown" is returned

Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [x] There is a single JIRA ticket related to the pull request. 
- [x] The web-link to the pull request is attached to the JIRA ticket.
- [x] The JIRA ticket has the _Patch Available_ state.
- [x] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [x] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheGroupView.java [hunk 1]
```diff
@@ -150,4 +152,21 @@ public int rebalanceOrder() {
     public int backups() {
         return ccfg.getBackups();
     }
+
+    /** @return {@code Yes} if group has expired entries, {@code No} otherwise. If {@code eagerTtl = true} returns 'Unknown'*/
+    public String hasEntriesPendingExpire() {
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheGroupView.java]: Removed it

File: modules/core/src/test/java/org/apache/ignite/internal/metric/SystemViewCacheExpiryPolicyTest.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            {-1, 0, -1, "Eternal"},
+            {0, 1, -1, "[update=1MILLISECONDS]"}
+        });
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTestsStarted() throws Exception {
+        super.beforeTestsStarted();
+
+        cleanPersistenceDir();
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/core/src/test/java/org/apache/ignite/internal/metric/SystemViewCacheExpiryPolicyTest.java]: Removed them

File: modules/core/src/test/java/org/apache/ignite/internal/metric/SystemViewSelfTest.java [hunk 3]
```diff
@@ -259,6 +264,215 @@ public void testCacheGroupsView() throws Exception {
...
+    @Test
+    public void testCacheViewExpiryPolicyWithPersist() throws Exception {
+        testCacheViewExpiryPolicy(true);
+    }
+
+    /** Tests work of {@link SystemView} for cache groups expiry policy info */
+    private void testCacheViewExpiryPolicy(boolean withPersistance) throws Exception {
+        try (IgniteEx g = !withPersistance ? startGrid() : startGrid(getConfiguration().setDataStorageConfiguration(
+            new DataStorageConfiguration().setDefaultDataRegionConfiguration(
+                new DataRegionConfiguration().setPersistenceEnabled(true)
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/core/src/test/java/org/apache/ignite/internal/metric/SystemViewSelfTest.java:282]: I don't think so. There is a ternanry conditional operator inside the try

`IgniteEx g = !withPersistance ? startGrid() : startGrid(getConfiguration()...`

Nevertheless, I checked the configuration for each case (`withPersistance = true/false`), the falg is working correctly

File: modules/core/src/test/java/org/apache/ignite/internal/metric/SystemViewSelfTest.java [hunk 4]
```diff
@@ -259,6 +264,215 @@ public void testCacheGroupsView() throws Exception {
         }
     }
 
+    /** Tests work of {@link SystemView} for cache expiry policy info with in-memory configuration */
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/core/src/test/java/org/apache/ignite/internal/metric/SystemViewSelfTest.java]: Fixed the typo

File: modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java [hunk 5]
```diff
@@ -331,7 +335,46 @@ public String cacheWriterFactory() {
 
     /** @see CacheConfiguration#getExpiryPolicyFactory() */
     public String expiryPolicyFactory() {
-        return toStringSafe(cache.cacheConfiguration().getExpiryPolicyFactory());
+        if (cache.cacheConfiguration().getExpiryPolicyFactory() == null)
+            return null;
+
+        ExpiryPolicy expiryPlc = (ExpiryPolicy)cache.cacheConfiguration().getExpiryPolicyFactory().create();
+
+        String expiryPlcFactoryStr = "";
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java]: Refactored it

File: modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java [hunk 6]
```diff
@@ -331,7 +335,46 @@ public String cacheWriterFactory() {
...
+        catch (IgniteCheckedException e) {
+            return e.getMessage();
+        }
+    }
+
+    /**
+     * Returns {@link Duration} String representation with description
+     * @param msg descriptive message for {@link Duration} instance
+     * @param duration @link Duration}
+     * @return {@link Duration} representation with descriptive message
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java]: Fixed the typo

File: modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java [hunk 7]
```diff
@@ -331,7 +335,46 @@ public String cacheWriterFactory() {
...
+
+        ExpiryPolicy expiryPlc = (ExpiryPolicy)cache.cacheConfiguration().getExpiryPolicyFactory().create();
+
+        String expiryPlcFactoryStr = "";
+
+        expiryPlcFactoryStr += durationToStringWithCustomMessage("create", expiryPlc.getExpiryForCreation());
+        expiryPlcFactoryStr += durationToStringWithCustomMessage("update", expiryPlc.getExpiryForUpdate());
+        expiryPlcFactoryStr += durationToStringWithCustomMessage("access", expiryPlc.getExpiryForAccess());
+
+        return expiryPlcFactoryStr.equals("") ? "Eternal" : expiryPlcFactoryStr;
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java]: There might be no custom toString() method from the user for this factory. If so, ClassName@HexHash might become a problem. It was actually the reason why we wanted changes here in the first place.

In addition I can add a className factory to the output. Usually it's incorporates the logic behind it. But it's only assumption. In general it's strange to give the user information about how the develoer decided to name the class. 

Maybe there is other way, but I thought ttl information should suffice. It's short, readable and it contains everything the user want to know about the factory

WGYT?

File: modules/core/src/test/java/org/apache/ignite/internal/metric/SystemViewCacheExpiryPolicyTest.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+     * Test for {@link CacheView} expiry policy factory representation. The test initializes the {@link CacheConfiguration}
+     * with custom {@link PlatformExpiryPolicyFactory}. Given different ttl input, the test checks the {@link String}
+     * expiry policy factory outcome for {@link CacheView#expiryPolicyFactory()}.
+     */
+    @Test
+    public void testCacheViewExpiryPolicy() throws Exception {
+        try (IgniteEx g = startGrid()) {
+            CacheConfiguration<Integer, Integer> ccfg = new CacheConfiguration<>();
+            ccfg.setName("cache");
+            ccfg.setExpiryPolicyFactory(new PlatformExpiryPolicyFactory(create, update, access));
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/core/src/test/java/org/apache/ignite/internal/metric/SystemViewCacheExpiryPolicyTest.java]: Added other factories to the test

File: modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java [hunk 9]
```diff
@@ -331,7 +336,46 @@ public String cacheWriterFactory() {
 
     /** @see CacheConfiguration#getExpiryPolicyFactory() */
     public String expiryPolicyFactory() {
-        return toStringSafe(cache.cacheConfiguration().getExpiryPolicyFactory());
+        if (cache.cacheConfiguration().getExpiryPolicyFactory() == null)
+            return null;
+
+        ExpiryPolicy expiryPlc = (ExpiryPolicy)cache.cacheConfiguration().getExpiryPolicyFactory().create();
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java:342]: We should also expose information about factory (according to method name)

File: modules/core/src/test/java/org/apache/ignite/internal/metric/SystemViewCacheExpiryPolicyTest.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+import static org.apache.ignite.internal.processors.cache.ClusterCachesInfo.CACHES_VIEW;
+
+/** Tests for {@link CacheView} expiry policy factory representation. */
+@RunWith(Parameterized.class)
+public class SystemViewCacheExpiryPolicyTest extends GridCommonAbstractTest {
+    /** {@link Factory} instances for test with different expiry policy. */
+    @SuppressWarnings("unchecked")
+    private static final Factory[] TTL_FACTORIES = {
+        null,
+        new FactoryBuilder.SingletonFactory<ExpiryPolicy>(new EternalExpiryPolicy()),
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/metric/SystemViewCacheExpiryPolicyTest.java:49]: `SingletonFactory` created via `EternalExpiryPolicy.factoryOf` methods

File: modules/core/src/test/java/org/apache/ignite/internal/metric/SystemViewSelfTest.java [hunk 11]
```diff
@@ -259,6 +263,114 @@ public void testCacheGroupsView() throws Exception {
...
+
+                        break;
+
+                    case "eagerTtlCache":
+                        assertEquals("Unknown", row.expiryPolicyEntry());
+
+                        break;
+
+                    case "withoutGrpCache":
+                        assertEquals("There is no cache group related data", row.expiryPolicyEntry());
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/metric/SystemViewSelfTest.java:366]: It's not correct, here cache group is created with `sharedGroup == false` flag.

File: modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheGroupView.java [hunk 12]
```diff
@@ -18,6 +18,7 @@
 package org.apache.ignite.spi.systemview.view;
 
 import java.util.Map;
+import org.apache.ignite.IgniteCheckedException;
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheGroupView.java]: Fixed it

File: modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheGroupView.java [hunk 13]
```diff
@@ -26,6 +27,7 @@
 import org.apache.ignite.configuration.CacheConfiguration;
 import org.apache.ignite.configuration.TopologyValidator;
 import org.apache.ignite.internal.managers.systemview.walker.Order;
+import org.apache.ignite.internal.processors.cache.CacheGroupContext;
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheGroupView.java]: Fixed it

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheOffheapManagerImpl.java [hunk 14]
```diff
@@ -1180,6 +1180,18 @@ private int expireInternal(
...
     }
 
+    /** {@inheritDoc} */
+    @Override public boolean hasEntriesPendingExpire(int cacheId) throws IgniteCheckedException {
+        if (pendingEntries == null)
+            return false;
+
+        PendingRow row = new PendingRow(cacheId);
+
+        GridCursor<PendingRow> cursor = pendingEntries.find(row, row, PendingEntriesTree.WITHOUT_KEY);
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheOffheapManagerImpl.java]: Fixed it

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java [hunk 15]
```diff
@@ -2691,6 +2700,27 @@ public long expiredSize() throws IgniteCheckedException {
...
+            if (pendingTree == null)
+                return false;
+
+            PendingRow row = new PendingRow(cacheId);
+            GridCursor<PendingRow> cur;
+
+            if (grp.sharedGroup())
+                cur = pendingTree.find(row, row, PendingEntriesTree.WITHOUT_KEY);
+            else
+                return init0(true) != null && !pendingTree.isEmpty();
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java]: Fixed it

File: modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java [hunk 16]
```diff
@@ -331,7 +336,46 @@ public String cacheWriterFactory() {
...
+
+        expiryPlcFactoryStrBld.append(func.apply("create", expiryPlc.getExpiryForCreation()));
+        expiryPlcFactoryStrBld.append(func.apply("update", expiryPlc.getExpiryForUpdate()));
+        expiryPlcFactoryStrBld.append(func.apply("access", expiryPlc.getExpiryForAccess()));
+
+        return expiryPlcFactoryStrBld.length() == 0 ? "Eternal" : expiryPlcFactoryStrBld.toString();
+    }
+
+    /** @return {@code Yes} if cache has expired entries, {@code No} otherwise. If {@code eagerTtl = true} returns 'Unknown'. */
+    public String expiryPolicyEntry() {
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java]: Yep. It's more informative

File: modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java (hunk 9 above)
Comment (maksaska) [reply modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java:345]: Added information about the factory as well

File: modules/core/src/test/java/org/apache/ignite/internal/metric/SystemViewCacheExpiryPolicyTest.java (hunk 10 above)
Comment (maksaska) [reply modules/core/src/test/java/org/apache/ignite/internal/metric/SystemViewCacheExpiryPolicyTest.java:48]: Added additional test cases

File: modules/core/src/test/java/org/apache/ignite/internal/metric/SystemViewSelfTest.java (hunk 11 above)
Comment (maksaska) [reply modules/core/src/test/java/org/apache/ignite/internal/metric/SystemViewSelfTest.java]: Fixed it

File: modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java [hunk 17]
```diff
@@ -331,7 +339,46 @@ public String cacheWriterFactory() {
...
+
+        Function<Duration, String> func = (duration) -> {
+            if (duration == null)
+                return null;
+
+            return new StringBuilder()
+                .append(duration.getDurationAmount())
+                .append(' ')
+                .append(duration.getTimeUnit() == null ? TimeUnit.MILLISECONDS : duration.getTimeUnit())
+                .toString();
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java:355]: Here is string still created, so StringBuilder is not required.
Let's handle ETERNAL somehow (compare with Duration.ETERNAL) and omit `null` components. For example: `CreatedExpiryPolicy [create=100 MILLISECONDS]`,  `EternalExpiryPolicy [create=ETERNAL]`,  etc

File: modules/core/src/test/java/org/apache/ignite/internal/metric/SystemViewCacheExpiryPolicyTest.java (hunk 10 above)
Comment (maksaska) [reply modules/core/src/test/java/org/apache/ignite/internal/metric/SystemViewCacheExpiryPolicyTest.java]: Removed additional cases. Fixed constructors: removed 'new' to EternalExpiryPolicy().factoryof()

File: modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java (hunk 17 above)
Comment (maksaska) [reply modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java]: Tried to reduce the number of Strings in favor of StringBuilder. Added checks for `ETERNAL` and `ZERO` durations. Refactored the logic for `ExpiryPolicy` string representation to take into account nulls

File: modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java [hunk 18]
```diff
@@ -344,26 +344,58 @@ public String expiryPolicyFactory() {
...
+     * @param out {@link StringBuilder} to append to.
+     * @param fieldName create/update/access expiry policy field name.
+     * @param durationSupplier Duration supplier for specified expiry policy field name.
+     * @param first {@link AtomicBoolean} flag indicating whether the field is the first in sequence.
+     */
+    private static void appendField(
+        StringBuilder out,
+        String fieldName,
+        Supplier<Duration> durationSupplier,
+        AtomicBoolean first) {
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java:373]: Put `) {` to the next line

File: modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java [hunk 19]
```diff
@@ -344,26 +344,58 @@ public String expiryPolicyFactory() {
...
+    /**
+     * @param out {@link StringBuilder} to append to.
+     * @param fieldName create/update/access expiry policy field name.
+     * @param durationSupplier Duration supplier for specified expiry policy field name.
+     * @param first {@link AtomicBoolean} flag indicating whether the field is the first in sequence.
+     */
+    private static void appendField(
+        StringBuilder out,
+        String fieldName,
+        Supplier<Duration> durationSupplier,
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java:372]: `durationSupplier` is senseless, just use duration.

File: modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java (hunk 18 above)
Comment (maksaska) [reply modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java]: Fixed

File: modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java (hunk 19 above)
Comment (maksaska) [reply modules/core/src/main/java/org/apache/ignite/spi/systemview/view/CacheView.java]: Done!

**Review by alex-plekhanov — APPROVED** [review]

**Pull request closed without merging**